src/app/topic/topic.service.ts

```diff
@@ -1,66 +1,122 @@
-        
-        },
-      ]
-    };
-  
-  topicDone = {
-    1: false,
-    2: false
-  }
+  topic :Topic;
+  topicProgress: number;
   
-  constructor() { }
-  getTopic(id): Observable<any> {
-    return of(this.topic);
+  constructor(
+    private request: RequestService,
+    private utils: UtilsService,
+  ) { }
+
+  getTopic(id: number): Observable<any> {
+    return this.request.get(api.get.stories, {params: { model_id: id }})
+      .pipe(map(response => {
+        if (response.success && response.data) {
+          return this._normaliseTopic(response.data);
+        }
+      })
+    );
   }
-  saveTopicRead(topicId) {
-    console.log('topic is marked as read.');
+
+  private _normaliseTopic (data: any){
+     // In API response, 'data' is an array of topics(since we passed topic id, it will return only one topic, but still in array format). That's why we use data[0]
+     if (!Array.isArray(data) || !this.utils.has(data[0], 'Story') || !this.utils.has(data[0], 'Filestore')) {
+      return this.request.apiResponseFormatError('Story format error');
+    }
+
+    let topic: Topic = {
+      id: 0,
+      programId: 0,
+      title: '',
+      content: '',
+      videolink: '',
+      hasComments: false,
+      files:[]
+    };
+    let thisTopic = data[0];
+      if (!this.utils.has(thisTopic.Story, 'id') || 
+          !this.utils.has(thisTopic.Story, 'title'))
+        return this.request.apiResponseFormatError('Story.Story format error');
+      
+    topic.id = thisTopic.Story.id;
+    if (this.utils.has(thisTopic.Story, 'program_id')) {
+      topic.programId = thisTopic.program_id;
+    }
+    topic.title = thisTopic.Story.title;
+    if (this.utils.has(thisTopic.Story, 'content')) {
+      topic.content = thisTopic.Story.content;
+    }
+    if (this.utils.has(thisTopic.Story, 'videolink')) {
+      topic.videolink = thisTopic.Story.videolink;
+    }
+    topic.hasComments = thisTopic.Story.has_comments;
+    topic.files = thisTopic.Filestore.map(item => ({url:item.slug , name:item.name}));
+
+    return topic;
   }
-  getTopicIsDone(topicId) {
-    return of(this.topicDone[topicId] ? this.topicDone[topicId] : false);
+  
+  updateTopicStatus(id){
+    let postData;
+   
+    postData = {
+      model: "topic",
+      model_id: id,
+      state: "completed"
+    }
+    return this.request.post(api.post.updateProgress, postData);
   }
-}
+  
+  getTopicProgress(activityId, topicId): Observable<any> {
+    return this.request.get(api.get.progress, {params: {
+        model: 'Activity',
+        model_id: activityId,
+        scope: 'Task'
+      }})
+      .pipe(map(response => {
+        if (response.success && !this.utils.isEmpty(response.data)) {
+
+            var progress = response.data.Activity.Topic.find(function (topic) {
+                return topic.id === topicId;
+            }) 
+            this.topicProgress = progress.progress;
+            return this.topicProgress;
```

Comment:
If we only use `this.topicProgress` for the return the data, why do we need this variable?